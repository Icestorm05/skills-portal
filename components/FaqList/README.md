# FAQ List

## Functionality

The FAQ list component displays a list of questions as expansion panels.

### Props

`<faq-list>` supports the following custom component attributes:

| attribute | type | description
| --- | --- | ---
| `faqs` | Array | The array of question objects. Each question object should follow the faq schema.
| `lazy` | Boolean *optional* | Whether to render the expansion panels lazily or not. Defaults to `true`.

### Schemas

The `faqs` prop should contain an array of objects, which adhere to the following schema:

#### Question Schema

| attribute | type | description
| --- | --- | ---
| `question` | String | The question.
| `answers` | Array | The array of answers. Each answer should follow the FAQ Answer schema.

#### Answer Schema

| attribute | type | description
| --- | --- | ---
| `type` | String | The tag to display the answer as. Accepts either: `p`, `b`, `ul` or `table`.
| `text` | String OR Array | The answer or array of answers. If the type is `p` or `b`, a string is expected. If the type is `ul`, an array is expected. If the type is `table`, an array of arrays is expected.

### Example

```html
<faq-list :faqs="faqs">
```

```ts
import Vue from 'vue';
import Component from 'nuxt-class-component';
import FaqList from '~/components/FaqList/FaqList.vue';
@Component({
    components: {
        FaqList,
    },
})
export default class GenericComponent extends Vue {
    private faqs = [{
        question: 'Test Question',
        answers: [{
            type: 'b',
            text: 'Test Answer',
        }, {
            type: 'table',
            text: [['row1-col1', 'row1-col2'], ['row2-col1', 'row2-col2']],
        }, {
            type: 'ul',
            text: ['Item 1', 'Item 2'],
        }, {
            type: 'p',
            text: 'Test Paragraph',
        }],
    }]
}
```