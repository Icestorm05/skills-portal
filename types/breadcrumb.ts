import { Location } from 'vue-router';

export interface IBreadcrumb {
    link: Location;
    text: string;
}