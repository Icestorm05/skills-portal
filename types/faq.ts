export interface IFaq {
    question: string;
    answers: IFaqAnswer[];
}

export interface IFaqAnswer {
    type: 'p' | 'b' | 'ul' | 'table';
    text: string | string[] | string[][];
}