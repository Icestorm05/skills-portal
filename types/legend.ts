export interface ILegend {
    SkillCategories: ISkillCategoryLegend;
    Levels: ISkillLevelLegend;
    Currencies: ISkillCurrencyLegend;
    fallback: string;
}

export interface ISkillCategoryLegend {
    [key: string]: string;
}

export interface ISkillLevelLegend {
    Awareness: string;
    Practitioner: string;
    Expert: string;
}

export interface ISkillCurrencyLegend {
    Historic: string;
    Fading: string;
    Current: string;
}