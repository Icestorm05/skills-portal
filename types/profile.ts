export interface IProfile {
    Biography: string;
    Company: string;
    Director: string;
    EmailAddress: string;
    EmployeeId: number;
    FirstLineReporting: string;
    FullName: string;
    JobTitle: string;
    KnownAs: string;
    MainStaffNumber: string;
    SecondLineReporting: string;
    Surname: string;
    children: IProfile[];
}