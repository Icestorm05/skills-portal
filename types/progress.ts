export interface IProgress {
    display: boolean;
}