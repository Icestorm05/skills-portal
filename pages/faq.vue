<template lang="pug">
    #FaqPage.text-xs-center
        h2.primary--text.mb-3 Frequently Asked Questions
        faq-list(:faqs="faqs")
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { State } from "vuex-class";

import { IFaq } from "types/faq";

import FaqList from "~/components/FaqList/FaqList.vue";

@Component({
  components: {
    FaqList
  }
})
export default class FaqPage extends Vue {
  @State("faqs") private faqs: IFaq[];

  private mounted() {
    this.$store.commit("progress/hide");
  }
}
</script>

<style lang="stylus" scoped>
#FaqPage {
    width: 100%;
}
</style>
