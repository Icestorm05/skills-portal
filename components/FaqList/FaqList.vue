<template lang="pug">
    v-expansion-panel(expand
                      popout
                      v-if="faqs.length")#FaqList
        v-expansion-panel-content(v-for="faq in faqs"
                                  :key="faq.question"
                                  :lazy="lazy"
                                  ripple).FaqPanel.mb-2
            span(slot="header").white--text.FaqPanelHeader {{ faq.question }}
            v-card.FaqPanelCard
                template(v-for="(answer, i) in faq.answers")
                    v-card-title(:key="answer.text"
                                 v-if="answer.type === 'b'").FaqPanelCardContentTitle.py-0
                        b {{ answer.text }}
                    v-card-text(:key="answer.text"
                                v-if="answer.type === 'p'"
                                :class="{ 'pt-0': i !== 0 }").FaqPanelCardContent.pb-0
                        p {{ answer.text }}
                    v-card-text(v-if="answer.type === 'ul'"
                                :class="{ 'pt-0': i !== 0, 'pb-0': i === faq.answers.length-1 }").FaqPanelCardContent
                        ul.ml-3
                            li(v-for="text in answer.text") {{ text }}
                    v-data-table(v-if="answer.type === 'table'"
                                :items="answer.text"
                                hide-actions
                                hide-headers
                                :class="{ 'pb-3': i !== faq.answers.length-1 }").FaqPanelCardContent
                        template(slot="items" slot-scope="{ item: items }")
                            td(v-for="item in items") {{ item }}
</template>

<script lang="ts">
import Component from "nuxt-class-component";
import Vue from "vue";
import { Prop } from "vue-property-decorator";

import { IFaq } from "types/faq";

import VCard, { VCardText, VCardTitle } from "vuetify/es5/components/VCard";
import VDataTable from "vuetify/es5/components/VDataTable";
import VExpansionPanel, {
  VExpansionPanelContent
} from "vuetify/es5/components/VExpansionPanel";
import { VContainer } from "vuetify/es5/components/VGrid";

@Component({
  components: {
    VCard,
    VCardTitle,
    VCardText,
    VContainer,
    VDataTable,
    VExpansionPanel,
    VExpansionPanelContent
  }
})
export default class FaqList extends Vue {
  @Prop({ type: Array, required: true })
  private faqs: IFaq[];
  @Prop({ type: Boolean, default: true })
  private lazy: boolean;
}
</script>

<style lang="stylus" scoped>
theme = json('../../assets/json/theme.json', { hash: true });

.FaqPanel {
    &.expansion-panel__container--active {
        margin-top: 8px;
        margin-left: 0;
        margin-right: 0;
    }

    /deep/ .expansion-panel__header {
        background-color: theme.secondary;

        .icon {
            color: rgba(255, 255, 255, 0.54) !important;
        }
    }
}
</style>
