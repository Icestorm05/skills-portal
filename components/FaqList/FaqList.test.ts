import FaqList from "./FaqList.vue";

import { IFaq, IFaqAnswer } from "types/faq";

import { createRenderer } from "vue-server-renderer";
import { mount } from "@vue/test-utils";

const pQuestion = {
  question: "Test Question 1",
  answers: [
    { type: "p", text: "Test Paragraph 1" },
    { type: "p", text: "Test Paragraph 2" }
  ] as IFaqAnswer[]
};

const bQuestion = {
  question: "Test Question 2",
  answers: [
    { type: "b", text: "Test Bold Answer 1" },
    { type: "b", text: "Test Bold Answer 2" }
  ] as IFaqAnswer[]
};

const ulQuestion = {
  question: "Test Question 3",
  answers: [
    { type: "ul", text: ["Test Bullet Answer 1", "Test Bullet Answer 2"] }
  ] as IFaqAnswer[]
};

const tableQuestion = {
  question: "Test Question 4",
  answers: [
    {
      type: "table",
      text: [["row1-col1", "row1-col2"], ["row2-col1", "row2-col2"]]
    }
  ] as IFaqAnswer[]
};

const faqs: IFaq[] = [pQuestion, bQuestion, ulQuestion, tableQuestion];

declare const expect: jest.Expect;

describe("FAQ List", () => {
  test("is a Vue instance", () => {
    const wrapper = mount(FaqList, { propsData: { faqs } });
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test("snapshot has not changed", () => {
    const wrapper = mount(FaqList, { propsData: { faqs, lazy: false } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("lazy snapshot has not changed", () => {
    const wrapper = mount(FaqList, { propsData: { faqs, lazy: true } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("renders all questions as expansion headers", () => {
    const wrapper = mount(FaqList, { propsData: { faqs } });
    expect(
      wrapper
        .findAll(".FaqPanelHeader")
        .at(0)
        .text()
    ).toContain("Test Question 1");
    expect(
      wrapper
        .findAll(".FaqPanelHeader")
        .at(1)
        .text()
    ).toContain("Test Question 2");
    expect(
      wrapper
        .findAll(".FaqPanelHeader")
        .at(2)
        .text()
    ).toContain("Test Question 3");
    expect(
      wrapper
        .findAll(".FaqPanelHeader")
        .at(3)
        .text()
    ).toContain("Test Question 4");
  });

  test("renders nothing if blank array provided", () => {
    const wrapper = mount(FaqList, { propsData: { faqs: [] } });
    expect(wrapper.find("#FaqList").exists()).toBeFalsy();
  });

  test("does not render answers with lazy prop", () => {
    const wrapper = mount(FaqList, { propsData: { faqs, lazy: true } });
    expect(wrapper.find(".FaqPanelCardContent").exists()).toBeFalsy();
  });

  test("does render answers with no lazy prop", () => {
    const wrapper = mount(FaqList, { propsData: { faqs, lazy: false } });
    expect(wrapper.find(".FaqPanelCardContent").exists()).toBeTruthy();
  });

  test("renders p answers as paragraph tags", () => {
    const wrapper = mount(FaqList, {
      propsData: { faqs: [pQuestion], lazy: false }
    });
    const card = wrapper.find(".FaqPanelCard");
    expect(card.findAll("p").length).toEqual(2);
    expect(card.html()).toContain("Test Paragraph 1");
    expect(card.html()).toContain("Test Paragraph 2");
  });

  test("renders b answers as bold tags", () => {
    const wrapper = mount(FaqList, {
      propsData: { faqs: [bQuestion], lazy: false }
    });
    const card = wrapper.find(".FaqPanelCard");
    expect(card.findAll("b").length).toEqual(2);
    expect(card.html()).toContain("Test Bold Answer 1");
    expect(card.html()).toContain("Test Bold Answer 2");
  });

  test("renders ul answers as bold tags", () => {
    const wrapper = mount(FaqList, {
      propsData: { faqs: [ulQuestion], lazy: false }
    });
    const card = wrapper.find(".FaqPanelCard");
    expect(card.findAll("ul").length).toEqual(1);
    expect(card.findAll("li").length).toEqual(2);
    expect(card.html()).toContain("Test Bullet Answer 1");
    expect(card.html()).toContain("Test Bullet Answer 2");
  });

  test("renders table answers as table tags", () => {
    const wrapper = mount(FaqList, {
      propsData: { faqs: [tableQuestion], lazy: false }
    });
    const card = wrapper.find(".FaqPanelCard");
    expect(card.findAll("table").length).toEqual(1);
    expect(card.findAll("tr").length).toEqual(2);
    expect(card.findAll("td").length).toEqual(4);
    expect(
      card
        .findAll("tr")
        .at(0)
        .findAll("td")
        .at(0)
        .text()
    ).toEqual("row1-col1");
    expect(
      card
        .findAll("tr")
        .at(0)
        .findAll("td")
        .at(1)
        .text()
    ).toEqual("row1-col2");
    expect(
      card
        .findAll("tr")
        .at(1)
        .findAll("td")
        .at(0)
        .text()
    ).toEqual("row2-col1");
    expect(
      card
        .findAll("tr")
        .at(1)
        .findAll("td")
        .at(1)
        .text()
    ).toEqual("row2-col2");
  });
});
