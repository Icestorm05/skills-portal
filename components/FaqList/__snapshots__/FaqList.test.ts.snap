// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FAQ List lazy snapshot has not changed 1`] = `
<ul class="expansion-panel expansion-panel--popout" id="FaqList">
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 1</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;"></div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 2</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;"></div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 3</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;"></div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 4</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;"></div>
    </li>
</ul>
`;

exports[`FAQ List snapshot has not changed 1`] = `
<ul class="expansion-panel expansion-panel--popout" id="FaqList">
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 1</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;">
            <div class="FaqPanelCard card" style="height: auto;">
                <!---->
                <div class="card__text FaqPanelCardContent pb-0">
                    <p>Test Paragraph 1</p>
                </div>
                <!---->
                <!---->
                <!---->
                <div class="card__text FaqPanelCardContent pb-0 pt-0">
                    <p>Test Paragraph 2</p>
                </div>
                <!---->
                <!---->
            </div>
        </div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 2</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;">
            <div class="FaqPanelCard card" style="height: auto;">
                <div class="card__title FaqPanelCardContentTitle py-0">
                    <b>Test Bold Answer 1</b>
                </div>
                <!---->
                <!---->
                <!---->
                <div class="card__title FaqPanelCardContentTitle py-0">
                    <b>Test Bold Answer 2</b>
                </div>
                <!---->
                <!---->
                <!---->
            </div>
        </div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 3</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;">
            <div class="FaqPanelCard card" style="height: auto;">
                <!---->
                <!---->
                <div class="card__text FaqPanelCardContent pb-0">
                    <ul class="ml-3">
                        <li>Test Bullet Answer 1</li>
                        <li>Test Bullet Answer 2</li>
                    </ul>
                </div>
                <!---->
            </div>
        </div>
    </li>
    <li tabindex="0" class="expansion-panel__container FaqPanel mb-2">
        <div class="expansion-panel__header"><span class="white--text FaqPanelHeader">Test Question 4</span>
            <div class="header__icon">
                <i aria-hidden="true" class="icon material-icons">keyboard_arrow_down</i>
            </div>
        </div>
        <div class="expansion-panel__body" style="display: none;">
            <div class="FaqPanelCard card" style="height: auto;">
                <!---->
                <!---->
                <!---->
                <div class="FaqPanelCardContent">
                    <div class="table__overflow">
                        <table class="datatable table">
                            <tbody>
                                <tr>
                                    <td>row1-col1</td>
                                    <td>row1-col2</td>
                                </tr>
                                <tr>
                                    <td>row2-col1</td>
                                    <td>row2-col2</td>
                                </tr>
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        </div>
    </li>
</ul>
`;
